src/nft_contract/src/main.nr

```diff
@@ -195,94 +192,156 @@ pub contract NFT {
-        to: AztecAddress,
+    fn _store_commitment_in_storage(context: &mut PublicContext, commitment: PartialNFTNote) {
+        context.storage_write(commitment.commitment(), true);
+    }
+
+    #[contract_library_method]
+    fn _initialize_transfer_commitment(
         context: &mut PrivateContext,
-        storage: Storage<&mut PrivateContext>,
+        private_nfts: Map<AztecAddress, PrivateSet<NFTNote, &mut PrivateContext>, &mut PrivateContext>,
+        from: AztecAddress,
+        to: AztecAddress,
     ) -> PartialNFTNote {
-        // We create a commitment with unpopulated/zero token id for 'to'
         let commitment = NFTNote::partial(
             to,
-            storage.private_nfts.at(to).storage_slot,
+            private_nfts.at(to).storage_slot,
             context,
             to,
-            context.msg_sender(),
+            from,
         );
-
-        NFT::at(context.this_address())._store_nft_set_commitment(commitment).enqueue(context);
-
+        NFT::at(context.this_address()).store_commitment_in_storage_internal(commitment).enqueue(context);
         commitment
     }
 
     #[contract_library_method]
-    fn _finalize_transfer_to_private(
+    fn _update_private_token_owner(
+        context: &mut PrivateContext,
+        private_nft_set: PrivateSet<NFTNote, &mut PrivateContext>,
         from: AztecAddress,
+        to: AztecAddress,
         token_id: Field,
-        commitment: PartialNFTNote,
-        context: &mut PublicContext,
-        storage: Storage<&mut PublicContext>,
     ) {
-        let public_owners_storage = storage.public_owners.at(token_id);
-        assert(public_owners_storage.read().eq(from), "invalid NFT owner");
+        let new_note = NFTNote::new(token_id, to);
+        private_nft_set.insert(new_note).emit(encode_and_encrypt_note(context, to, from));
+    }
+
+    #[contract_library_method]
+    fn _remove_private_token_owner(
+        private_nft_set: PrivateSet<NFTNote, &mut PrivateContext>,
+        token_id: Field,
+    ) {
+        let notes = private_nft_set.pop_notes(NoteGetterOptions::new()
+            .select(NFTNote::properties().token_id, Comparator.EQ, token_id)
+            .set_limit(1));
+        assert(notes.len() == 1, "nft not found in private to public");
+    }
 
-        // Set the public NFT owner to zero
-        public_owners_storage.write(AztecAddress::zero());
+    #[contract_library_method]
+    fn _update_public_token_owner(
+        public_owners: Map<Field, PublicMutable<AztecAddress, &mut PublicContext>, &mut PublicContext>,
+        from: AztecAddress, 
+        to: AztecAddress,
+        token_id: Field,
+        from_private: bool,
+    ) {
+        let owner = public_owners.at(token_id).read();
+        if from_private {
+            assert(owner.eq(AztecAddress::zero()), "token has public owner");
+        } else {
+            assert(owner.eq(from), "caller is not owner");
+        }
+        public_owners.at(token_id).write(to);
+    }
```

Comment:
i think it should be better to split the logic in one `_update_public_token_owner` and `_update_private_token_owner` than having the `from_private` boolean, that doubles the size of the proof

Reply:
also wondering if we should `AztecAddress::zero()` the "owned privately" status, wdyt?